Release ROS 2 simulation with O3DE (#2520)
o3de/o3de.org: content/docs/user-guide/interactivity/robotics/deploying-simulation.md

With assets processed, you can build a monolithic build, which includes all Gems and most libraries linked statically. 
It also excludes large editor-only frameworks like Qt.
Refer to [Create a project game release layout](docs/user-guide/packaging/windows-release-builds/#create-a-project-game-release-layout) to learn more.

To perform a monolithic build, use the `install` target:
```bash
cmake -B build/linux_mono -S . -G "Ninja Multi-Config"  -DLY_MONOLITHIC_GAME=1
cmake --build build/linux_mono --target install --config release
```

In the install directory `${PROJECT_PATH}/install/bin/Linux/release/Monolithic/`, you can find a package containing all necessary binaries and archives to run your simulation in the cloud or on another computer.

```
# tree ${PROJECT_PATH}/install/bin/Linux/release/Monolithic/
.
├── Cache
│   └── linux
│       └── engine.pak
├── libVkLayer_khronos_validation.so
├── Ros2Project.GameLauncher
├── Ros2Project.ServerLauncher
├── Ros2Project.UnifiedLauncher
└── VkLayer_khronos_validation.json
```

# Headless simulation

Often simulation needs to be run headless (e.g., as a part of CI/CD pipeline).
This can be achieved by disabling RHI in the GameLauncher. 

> To run headless, start with `-rhi=null`.

Author: ```suggestion
To run headless, start with `-NullRenderer -rhi=null`.
```

> To run headless, start with `--rhi=null`.

Reviewer: 1. You dont need two dashes. -rhi=null will work.
2. -rhi=null just disables the RHI layer. Atom is still running the RPI layer which maybe unnecessary here. To disable the RPI and RHI layer you would want to pass -NullRenderer as a CLI param or you should set NullRenderer to be true within settings registry under path "/O3DE/Atom/RPI/Initialization"

Author: Thanks, adjusted use `-rhi=null`. Thanks for the tip with `-NullRenderer`. I am sorry, but this caused to Game launcher to crash in a headless environment.
I what about passing both arguments: `-NullRenderer -rhi=null`. It seems to do a trick.

> To run headless, start with `--rhi=null`.

Author: ```suggestion
To run headless, start with `-NullRenderer -rhi=null`.
```

```bash 

> ./Ros2Project.GameLauncher -rhi=null

Author: ```suggestion
./Ros2Project.GameLauncher -NullRenderer -rhi=null
```

> ./Ros2Project.GameLauncher --rhi=null

Author: ```suggestion
./Ros2Project.GameLauncher -NullRenderer -rhi=null
```

---
linkTitle: Deploying Simulation 
title: Deploying Simulation
description: Deploying a headless test tool with O3DE.
toc: true
weight: 520
---

## Overview

Robotic simulations are often deployed as part of a CI/CD pipeline. In such scenarios, running simulations with the Editor is not recommended.

## Deploying Release Build Game Launcher

For detailed instructions, please refer to the [Packaging and Release](docs/user-guide/packaging/) section. Deploying a robotic simulation is akin to deploying a game, but the result is an executable that has ROS 2 communication channels.

> ## Step-by-Step Project Template Release

Reviewer 2: I would suggest a bit of rewording here overall - this is an example using ROS 2 Project Template.

Maybe the best would be: 
"Now you need to do X, (..)" - explain in general terms.

"For our example, the commands would be (..)"

Author: Adjusted.

## Monolithic build

With assets processed, you can build a monolithic build, which includes all Gems and most libraries linked statically. 
It also excludes large editor-only frameworks like Qt.
Refer to [Create a project game release layout](docs/user-guide/packaging/windows-release-builds/#create-a-project-game-release-layout) to learn more.

To perform a monolithic build, use the `install` target:
```bash
cmake -B build/linux_mono -S . -G "Ninja Multi-Config"  -DLY_MONOLITHIC_GAME=1
cmake --build build/linux_mono --target install --config release
```

In the install directory `${PROJECT_PATH}/install/bin/Linux/release/Monolithic/`, you can find a package containing all necessary binaries and archives to run your simulation in the cloud or on another computer.

```
# tree ${PROJECT_PATH}/install/bin/Linux/release/Monolithic/
.
├── Cache
│   └── linux
│       └── engine.pak
├── libVkLayer_khronos_validation.so
├── Ros2Project.GameLauncher
├── Ros2Project.ServerLauncher
├── Ros2Project.UnifiedLauncher
└── VkLayer_khronos_validation.json

```

# Headless simulation

> Often simulation need to be run headless (e.g., as a part of CI/CD pipeline).

Reviewer 2: ```suggestion
Often simulation needs to be run headless (e.g., as a part of CI/CD pipeline).
```

This can be achieved by disabling RHI in the GameLauncher. 
To run headless, start with `--rhi=null`.
```bash 
./Ros2Project.GameLauncher --rhi=null
```
The above command should work for environments that have no GPU. 

> If your simulation relies on the ROS2 Camera component, you will need RHI. 

Reviewer 2: ```suggestion
If your simulation relies on the ROS 2 Camera component, you will need RHI. 
```
